### ISSUE TITLE
opt: incorrect execbuilder configuration of required orderings when optional columns are present

### ISSUE THREAD

**Issue by RaduBerinde** [cockroachdb/cockroach#31882]
Consider the query:
```sql
CREATE TABLE abcd (a INT, b INT, c INT, d INT, INDEX abc(a, b, c) STORING (d))
SELECT a, b, ARRAY_AGG(d) AS s FROM (SELECT * FROM abcd ORDER BY c) GROUP BY a, b
```
GroupBy requires ordering `+c opt(a,b)` and the index can provide this ordering:
```
group-by
 ├── columns: a:1(int) b:2(int) s:6(int[])
 ├── grouping columns: a:1(int) b:2(int)
 ├── internal-ordering: +3 opt(1,2)
 ├── stats: [rows=1000, distinct(1,2)=1000, null(1,2)=19.9]
 ├── cost: 1130
 ├── key: (1,2)
 ├── fd: (1,2)-->(6)
 ├── prune: (6)
 ├── scan abcd@ab
 │    ├── columns: a:1(int) b:2(int) c:3(int) d:4(int)
 │    ├── stats: [rows=1000, distinct(1,2)=1000, null(1,2)=19.9]
 │    ├── cost: 1090
 │    ├── ordering: +3 opt(1,2)
 │    └── prune: (1-4)
 └── aggregations
      └── array-agg [type=int[], outer=(4)]
           └── variable: d [type=int]
```

The problem is when we configure the required ordering of the scan:
```
group           ·            ·             (a, b, s)     ·
 │              aggregate 0  a             ·             ·
 │              aggregate 1  b             ·             ·
 │              aggregate 2  array_agg(d)  ·             ·
 │              group by     @1-@2         ·             ·
 └── render     ·            ·             (a, b, d)     ·
      │         render 0     a             ·             ·
      │         render 1     b             ·             ·
      │         render 2     d             ·             ·
      └── scan  ·            ·             (a, b, c, d)  +c
·               table        abcd@ab       ·             ·
·               spans        ALL           ·             ·
```
The index can cannot actually provide `+c`; it can provide `+a,+b,+c`. This information is used to configure distsql input synchronizers so this can result in incorrect results and/or `incorrectly ordered stream` errors.

**Comment by RaduBerinde** [comment]
It's easy enough to fix this for scans, but I don't have a good idea of how to handle this in general. We need a way to obtain a specific `Ordering` for each node in the tree.

**Comment by RaduBerinde** [comment]
CC @andy-kimball

**Comment by RaduBerinde** [comment]
The fix requires plumbing the "provided" ordering of each operator (which is a subset of the "required" ordering). This is a relatively big change and the code will be pretty different in the new representation (making the backport difficult).

I am considering a temporary fix for 2.1.x - don't set grouping columns as optional columns in Group-By. AFAIK this is the only case where we have optional columns that are not due to FDs (i.e. constant columns, which don't cause the issue here). This would be done together with adding an exploration rule for `GroupBy` which tries to use interesting orderings (#30000). This should cover most streaming aggregation opportunities that we would otherwise lose with the fix (and cover many new cases).

**Comment by RaduBerinde** [comment]
Another example of where optional problems can cause issues is under a filter that constraints a column to a constant. In the example below the scan is incorrectly configured with the `+a,+c` ordering. Fortunately this does not cause issues in practice because the distsql planner always pushes the filter down into each node.

```
root@127.77.95.175:35097/defaultdb> EXPLAIN (VERBOSE) SELECT * FROM (SELECT a, b, c FROM abc LIMIT 10) WHERE b = 1 ORDER BY a, b, c;
    tree    | field  |    description    |  columns  | ordering  
+-----------+--------+-------------------+-----------+----------+
  filter    |        |                   | (a, b, c) | +a,+c     
   │        | filter | b = 1             |           |           
   └── scan |        |                   | (a, b, c) | +a,+c     
            | table  | abc@abc_a_b_c_idx |           |           
            | spans  | ALL               |           |           
            | limit  | 10                |           |           
```
